core/src/main/scala/kafka/utils/ZkUtils.scala

```diff
@@ -892,3 +871,219 @@ object ZkPath {
     client.createPersistentSequential(path, data)
   }
 }
+
+/**
+ * Implements an ephemeral znode that is recreated in the case it disappears.
+ * A znode can be deleted in this case if it has been created in a previous session.
+ * Say that the client that created it crashes leaving the session open. In this case,
+ * this client will try to create it and fail, but eventually the ephemeral znode will
+ * be remove and needs to be created and associated to this session.
+ *
+ * It additionally creates the parent path recursively.
+ */
+class ZKWatchedEphemeral(path : String,
+                          data : String,
+                          expectedCallerData: Any,
+                          checker: (String, Any) => Boolean,
+                          zkHandle : ZooKeeper) extends Logging {
+  private val createCallback = new CreateCallback
+  private val ephemeralWatcher = new EphemeralWatcher
+  private val existsCallback = new ExistsCallback
+  private val getDataCallback = new GetDataCallback
+  val latch: CountDownLatch = new CountDownLatch(1)
+  var result: Code = Code.OK
+  @volatile
+  private var stop : Boolean = false
+  private class CreateCallback extends StringCallback {
+    def processResult(rc : Int,
+                      path : String,
+                      ctx : Object,
+                      name : String) {
+      Code.get(rc) match {
+        case Code.OK => {
+          // check that exists and wait
+           checkAndWatch
```

Comment:
Why do we still need to trigger a ZooKeeper.exist with ExistsCallback again for this case?